### ISSUE TITLE
linkedlist для особо одарённых :)

### ISSUE THREAD

**Issue by uliss4ever** [uliss4ever/My_repository_2#7]
Лёш, вот prev_node и next_node - они существуют так же, как и new_node и old_node, вещи одного порядка? 
Вот у нас есть (будет) этот связный список. a -> b -> c -> d Мы аппендом хотим сделать что? Добавить после d - e? А где в таком случае значения, которые они хранят?
 Если ты не против, можно пока писать тут, а то я в письмах то одну мысль теряю, то другую. Сейчас я соберу их все, постараюсь)

**Comment by uliss4ever** [comment]
>Потому что _для какой_ ноды new_node является следующей (стоит
справа от неё), а для какой - предыдущей (стоит слева от неё)?

node_a -> new_node -> node_b
new_node следующая для а и предыдущая для b
Я там в письме не очень точно спросила. Надо ли определять next_node и prev_node сами по себе. То есть это вообще некорректный вопрос. У нас next и prev толкьо по отношению к чему-то, а не сами по себе, так?

**Comment by uliss4ever** [comment]
И вот тут, почему у b_node индекс - index ? мне почему-то казалось, что у неё должен стать index+1 а у new_node - индекс = index
Вот, я начала по одной:
new_node.prev_node - это а_node
old_node - это b_node
Только вот зачем нам задавать что такое a_node ? Может сразу связи? Потому что она как была а своём метсе, так и осталась. Двигается только old_node по-моему?

**Comment by Lex-2008** [comment]
> У нас next и prev толкьо по отношению к чему-то, а не сами по себе, так?

Да, правильно, именно так. Просто `next_node` не бывает, есть только `<нода>.next_node`.

> почему у b_node индекс - index ? мне почему-то казалось, что у неё должен стать index+1 а у new_node - индекс = index

Правильно, _должен стать_. Но **до** вставки у b_node индекс - index. А поскольку мы _сначала_ определяем a_node и b_node, а _потом_ вставляем новую, то должны использовать нумерацию до вставки :)

> Вот, я начала по одной:
> new_node.prev_node - это а_node

да, почти. Так _должно_ стать _после_ вставки. _До_ вставки, когда мы только вызвали insert - он создал новый элемент new_node, который сразу после создания не знает, кто должен стоять слева или справа от него. Поэтому мы должны сказать ему: "дорогой new_node, слева от тебя стоит вот этот нод, справа от тебя стоит вот этот нод", и двум другим нодам тоже. Одному: "дорогой нод, слева от тебя теперь стоит вот этот new_node", а другому - что справа.

**Comment by uliss4ever** [comment]
![Снимок экрана 2021-03-06 в 17 45 44](https://user-images.githubusercontent.com/32711430/110210679-d435d980-7ea3-11eb-9b1d-cf0d1e6b1830.png)

У меня появились новые мысли... Но пришлось выдумать новую переменную(?) old_node_left. Это после того как я нарисовала их все.

**Comment by Lex-2008** [comment]
Да, почти правильно. Вот это old_node_left и есть a_node. Единственное - у тебя сейчас пропадает та нода, что стояла между ними.

Давай с фруктами попробуем. Пусть у нас есть список:
[арбуз]<->[банан]<->[вишня]<->[груша]
и надо сделать insert(2, виноград). То есть, `index==2`, `new_node.data=виноград`.
На ноды с какими фруктами будут указывать old_node и old_node_left после выполнения первых трех строчек?
```
old_node = self._get_node_(index+1)
old_node_left = self._get_node_(index-1)
```
`index==2`, значит, `index+1==3`, значит old_node указывает на ноду с фруктом...
`index==2`, значит, `index-1==1`, значит old_node_left указывает на ноду с фруктом...

Потом мы говорим, что между ними стоит "виноград", и что у нас получается?

**Comment by uliss4ever** [comment]
>Единственное - у тебя сейчас пропадает та нода, что стояла между ними.

Погоди, так между ними же new_node теперь стоит? Про её индекс тоже надо было сказать? У неё инедкс = index

>[арбуз]<->[банан]<->[вишня]<->[груша]
и надо сделать insert(2, виноград). То есть, index==2, new_node.data=виноград
index==2, значит, index+1==3, значит old_node указывает на ноду [вишня]
index==2, значит, index-1==1, значит old_node_left указывает на ноду [банан]

**Comment by Lex-2008** [comment]
Нет. Давай ещё раз :)
Есть список из 4 нод:
[арбуз]<->[банан]<->[вишня]<->[груша]
какие тут номера у нод?
Какая нода номер 0?
Какая нода номер 1?
Какая нода номер 2?
Какая нода номер 3?

**Comment by uliss4ever** [comment]
Так подряд же вроде?
Какая нода номер 0? 
[арбуз]
Какая нода номер 1?
[банан]
Какая нода номер 2?
[вишня]
Какая нода номер 3?
[груша]
Вставили на место 2 виноград - то, что справа, уехало на 1 вправо.
То есть может я глобально что-то наоборот понимаю...

![IMG_8338](https://user-images.githubusercontent.com/32711430/110211958-20841800-7eaa-11eb-9b2d-b6ca6ba85efc.jpeg)

**Comment by Lex-2008** [comment]
Да, правильно. Только вот это:

> Вставили на место 2 виноград - то, что справа, уехало на 1 вправо.

это будет только _после_ того, как insert завершится. То есть, это **мы** должны сделать так, чтобы "то, что справа, уехало на 1 вправо".

Вначале инсёрта надо работать с тем, что было до запуска инсёрта - то, что у тебя нарисовано наверху.



Кстати, ты знаешь, откуда эти номера берутся? Ведь ноды не знают, под каким номером они идут! Это ведь когда мы их в `_get_node_` перебираем по очереди, переходя по ссылкам `current_node.next_node`, то считаем: это нулевая нода, это первая нода, и т.д. Вот в этих двух функциях:

```
    def _get_node_(self, index):
        for i, node in enumerate(self._node_iter()):
            if i == index:
                return node

    def _node_iter(self):
        current_node = self.head
        while current_node is not None:
            yield current_node
            current_node = current_node.next_node
```

`_node_iter` нам возвращает по очереди [арбуз], [банан], [вишня], [груша] - только потому что они связаны ссылками next_node.

`enumerate(self._node_iter())` добавляет им номера:
(0, [арбуз]),
(1, [банан]),
(2, [вишня]),
(3, [груша])

и только именно поэтому у нас нода [вишня] имеет номер 2.

**Comment by uliss4ever** [comment]
Ааа. Дошло. Тогда у нас получается old_node_left - 2-я нода, а old_node - это на текущий момент 3-я, то есть они подряд. Я совершенно уверена была, что они старой_нодой и старой_нодой_слева становятся только после insert  
Тогда:
old_node = self._get_node_(index) - должно быть в первой строке. То есть это мы должны интерпретатору рассказать, что у нас является на данный момент старой нодой, а не он нам...
Про остальное - щас...

**Comment by Lex-2008** [comment]
Точно!

Вот ты нарисовала две картинки: "до" и "после". И функция инсерт должна сделать из "до" в "после".

Тем, что ноды могут сдвигаться, можно пренебречь: представь, что это кусочки фруктов, которые плавают в супе, и единственное, что их связывает - это эти ниточки-ссылочки.

Номерами тоже можно пренебречь - это их enumerate проставляет.

Чем тогда две картинки отличаются? Ну только этими стрелочками!

Единственное - в то время, _пока_ мы меняем эти ссылочки-ниточки (то есть между первым и последним присваиваниями), функцию `self._get_node_(index)` вызывать нельзя, т.к. она переходит по ним и пока мы их не сделаем окончательно правильно, она может сломаться.

**Comment by uliss4ever** [comment]
А как же мне тогда быть? Я же сперва как раз задаю старую ноду с помощью этого метода. Или потом, в 3 строке он уже про old_node всё знает и _get_node_(index) не пользуется? 

![Снимок экрана 2021-03-06 в 19 20 37](https://user-images.githubusercontent.com/32711430/110213493-5973bb00-7eb1-11eb-8372-ad504756251c.png)

**Comment by Lex-2008** [comment]
Да, вот теперь всё правильно!

Да, `_get_node_` у тебя вызывается в первых двух строчках - она использует имеющиеся ссылки, находит нужные ноды и мы их сохраняем в `old_node` и `old_node_left`. Теперь эти ноды не потеряются, даже если `_get_node_` нельзя будет вызвать.

И после этого в 4 строчках ты совершенно правильно задаешь ссылки.

И где-то ещё внизу у нас есть `self._size += 1`, да?

И, кстати, после строчки `self.append(data)` на твоем скриншоте надо добавить строчку `return`.

Потому что `self.append()` добавляет одну ноду и делает `self._size += 1`, а если после этого ещё и вызвавший его инсёрт сделает `self._size += 1`, то в результате у нас добавится 1 нода, а `self._size` увеличится на 2 - и произойдёт :bug:

**Comment by uliss4ever** [comment]
Точно, про return забыла. А вот с size как раз сижу и думаю. У меня он вот где, сверху. И во идее он на всё распространяется же? На всё, кроме append. Или его вниз надо?
Не нравится он мне тут наверху. Значит, спева +1, потом проверяем, равен индекс 0 или нет, и дальше идём в аппенд. И если вдруг его отработали, то успели ещё +1 сделать. 

![Снимок экрана 2021-03-06 в 19 40 48](https://user-images.githubusercontent.com/32711430/110214116-241c9c80-7eb4-11eb-98d1-446ebc7c13b0.png)

**Comment by Lex-2008** [comment]
Да, давай-ка его лучше в самый низ, после последней строчки - иначе он `self._get_node_(index)` испортит.

И ещё в `elif` `- 1` не нужен :) Помнишь, я тебя спрашивал: если сделать `self.append(data)`, то какой номер получит новый элемент? Значит, `self.append(data)` эквивалентно `self.insert(что?, data)`?

**Comment by uliss4ever** [comment]
Ещё Енот сказал про инсёрт, что "можно по идее использовать родительский". После этого сразу стало ясно, что надо переопределить. Вот, как смогла...
И ещё - что переопределить self.head = None в clear. Вот как получилось. Теперь думаю над delite

![Снимок экрана 2021-03-06 в 19 40 58](https://user-images.githubusercontent.com/32711430/110214272-dfddcc00-7eb4-11eb-8d67-5379c2e6dfc7.png)

**Comment by uliss4ever** [comment]
> Да, давай-ка его лучше в самый низ, после последней строчки - иначе он `self._get_node_(index)` испортит.
> 
> И ещё в `elif` `- 1` не нужен :) Помнишь, я тебя спрашивал: если сделать `self.append(data)`, то какой номер получит новый элемент? Значит, `self.append(data)` эквивалентно `self.insert(что?, data)`?

Дааа, несколько раз хотела ответить, но в почте письма перепутались. Значит если у нас 15 нод, то size = 15. И если Мы делаем self.append(data), то индекс(data) после выполнения будет 15. И это эквивалентно  self.insert(15, data). Точно, -1 не нужен)

**Comment by Lex-2008** [comment]
> Ещё Енот сказал про инсёрт, что "можно по идее использовать родительский"

Хмм, я об этому думал, но мне показалось что так будет сложнее. Если хочешь - можем попробовать написать!

Да, по поводу clear - всё правильно, но с избытком. Смотри: *надо* сделать три вещи:
````
self._size = 0
self.head = None
self.tail = None
````
При этом, родительский метод, который мы вызываем через `super().clear()`, делает две вещи:
````
self._size = 0
self.head = None
````
Следовательно, если ты вызываешь `super().clear()`, то эти две вещи делать уже не надо - их за тебя сделает родительский метод. Следовательно, какую одну вещь осталось сделать?

**Comment by uliss4ever** [comment]
Ой!!! Не инсёрт - индекс! Я имела в виду def index(...) который после clear. И я его попыталась переопределить, там же с clear . Извини, пожалуйста. Мне самой новый инсёрт очень нравится!

**Comment by uliss4ever** [comment]
Clear - вот так тогда
![Снимок экрана 2021-03-06 в 20 10 03](https://user-images.githubusercontent.com/32711430/110214938-06e9cd00-7eb8-11eb-9327-b9963828a0e2.png)

**Comment by Lex-2008** [comment]
clear - правильно!

а index лучше оставить родительским, чем такое писать :)

У тебя в цикле for уже есть node, соответствующая индексу i. Зачем тебе в нем вызывать `self._get_node_(i)`? Совершенно не надо, а даже вредно, т.к. `self._get_node_(i)` пробегает по всему списку, помнишь об этом?

**Comment by uliss4ever** [comment]
Да... Я помню, просто хотела, чтобы он бегал либо с одной стороны, либо с другой. А это у нас заложено в get_node_. А в родительском он бежит с одного только конца
![Снимок экрана 2021-03-06 в 20 26 45](https://user-images.githubusercontent.com/32711430/110215416-bd4eb180-7eba-11eb-99d7-0827c3257ee4.png)



Совсем не так, да?

**Comment by Lex-2008** [comment]
Да, было бы хорошо, но так не получится. Потому что если у тебя есть 15 нод, и тебе нужна 12-я, то ты знаешь, с какой стороны к ней лучше подойти.

Но если у тебя есть 15 нод, и тебе нужно найти ту, где сидит корова Буренка? То с какой стороны надо идти? А кто её знает...

**Comment by uliss4ever** [comment]
Мда. То есть родительский тогда? Просто Енот так сказал, что было ощущение, что можно попытаться.
Я сделала некий delite... Все три штуки. Правда теперь боюсь, вдруг это будет последняя капля в том количестве тупизма, который я выдала сегодня.
В том варианте, где мы среднюю удаляем, есть две такие строчки:
 old_node.next_node = old_node.prev_node
 old_node.prev_node = old_node.next_node
 И они меня немного смущают.

**Comment by Lex-2008** [comment]
> ```
> old_node.next_node = old_node.prev_node
> old_node.prev_node = old_node.next_node
> ```

Да, это неправильно. Как минимум, потому что сначала мы присваиваем одно второму, а потом - второе первому. Ну, ты же понимаешь, почему вот это выглядит странно?

```
a = b # теперь a равно b
b = a # а теперь...
```

во-вторых, когда мы среднюю удаляем, то там надо соединить те, которые были с двух сторон от него, друг с другом. Ты картинку на это какую-нибудь нарисовала?

**Comment by uliss4ever** [comment]
Да, и у меня уже другие идеи возникли, пытаюсь перевести в код...

**Comment by uliss4ever** [comment]
Вот кажется
![Снимок экрана 2021-03-06 в 21 07 46](https://user-images.githubusercontent.com/32711430/110216391-0b19e880-7ec0-11eb-8872-f1157e4046b7.png)

**Comment by Lex-2008** [comment]
точно! правильно!! молодец!!!

А другие случаи покажешь?

**Comment by uliss4ever** [comment]
То есть я не совсем безнадежна?? (Если что, я на утешительную похвалу не напрашиваюсь, я с этого октября уже стооолько нового о своих умственных способностях узнала, что хуже уже не будет) 
Вот оно целиком, правда я пытаюсь понять, эти None мне точно нужны или они заложены в head и tail - они же потому голова и хвост, что у них одна из ссылок None
И да, мне никак не понять, почему мы всё-таки не можем делать отрицательную индексацию и здесь и в инсёрте?

![Снимок экрана 2021-03-06 в 21 15 36](https://user-images.githubusercontent.com/32711430/110216658-58e32080-7ec1-11eb-9552-63231c134173.png)

**Comment by uliss4ever** [comment]
Да, наверное всё-таки нельзя. Попыталась на встроенном списке - Питон запутался, не на то место поставил

![Снимок экрана 2021-03-06 в 21 27 36](https://user-images.githubusercontent.com/32711430/110217010-dc514180-7ec2-11eb-9555-a810ba1bec72.png)

**Comment by Lex-2008** [comment]
Да, всё правильно! Молодец!! Можно сдавать!!! На мои схемы смотрела или свою нарисовала?

Есть ещё небольшие оптимизации по производительности, и на твои вопросы я отвечу.

А, да, у встроенного питоновского списка "-1" значит "первый с конца, а точнее перед ним"

**Comment by uliss4ever** [comment]
Да ладно? Ура!! Я вопросы тогда тебе, наверное, на занятии задам, а то представляю как ты сейчас устал на это смотреть) Отправлю, и вдруг Енот что ещё скажет. И да, я за оптимизацию, обеими руками за!
По поводу картинок - я на твои схемы смотрела когда с insert разбиралась, и тогда уже delite принципиально поняла. Просто у меня реальный затык - мне было не догадаться что надо было в инсёрте ввести эту вторую старую ноду, так и здесь - что надо задать левую и правую. То есть я про них знала, но пыталась зачем-то через текущее переопределить. Когда догадалась как было надо - так просто стало казаться. Вот такое рисовала.
И это же начала вводить в коде, а потом передумала и получила это дурацкое a = b #  b = a #

![IMG_8339](https://user-images.githubusercontent.com/32711430/110217444-2a674480-7ec5-11eb-89fb-59097c1cd042.jpg)

Спасибо тебе огроменное. И за то, что подводил, подсказывал, и за то, что не подсказывал! Ты же знал, как мне эта self.old_node_left нужна, просто жизненно необходима?!
И за терпение... Я бы уже за это время убила кого-нибудь, а то чего они такие тупые))

**Comment by uliss4ever** [comment]
Вот мне брат однажды такое прислал и моё любимое - п.59 Прямо про меня один в один)
https://pikabu.ru/story/100_samyikh_yarkikh_tsitat_programmistov_3841765

**Comment by Lex-2008** [comment]
> self.old_node_left нужна

Ну да, я, вроде, тебе это с самого начала говорил... :) `old_node` и `prev_node`, потом `a_node` и `b_node`... Как это можно было понятнее сказать? `old_left` и `old_right`? :)

Да, там все цитаты хороши!

Вот ещё одной не хватает:
> Самое сложное в программировании - придумывать имена переменным.

**Comment by uliss4ever** [comment]
Не, что ты! Ты очень понятно объясняешь, правда. И всеми возможными способами. Просто я говорю - полный хаос в голове. Ну ок, на 85% И вот в эти проценты попало то, что я любой ценой решила обойтись без новой переменной. Но кстати да, именно на этапе a_node и b_node я начала что-то подозревать)))  
И да, если бы "ещё понятнее" - то это уже просто подсказать. А это не интересно совсем. 
Совсем не сложно было на 1 курсе, смотри: все функции - это def f(), все списки - это l, l1, l2.... Генераторы - g.

**Comment by Lex-2008** [comment]
Хмм, попробую настойчивей объяснять :)

Да, с хаосом надо как-то справляться, т.к. обычно в одном месте программы у нас примерно две-три переменные - только о них и надо думать :) Потом уже эти переменные на что-то указывают - на какие-то объекты - у объектов есть свои атрибуты, свойства, методы - это всё идёт во вторую очередь (через точку). И, как ты верно заметила (где-то на скриншотах видел), можно добавить ещё одну точку и получить атрибут атрибута... Например, `self.head` - это первая нода, а `self.head.next_node` - нода, следующая за первой, то есть, обычно, вторая. Ну или `None`, если в списке только одна нода. Или какой-нибудь AttributeError, если список вообще пуст - тогда `self.head` указывает на `None`, а у None нету атрибута `next_node`, и нам за попытку его вызвать прилетит AttributeError.

Да, `def f()` - это совсем нехорошее название функции. Ну, хорошо что в реальной жизни оно обычно не так! А как тебе нравится математика, где в каждом уравнении по иксу стоит?!

**Comment by uliss4ever** [comment]
>Хмм, попробую настойчивей объяснять :)

Как по мне - так идеально и так. Ну просто некоторые вещи до меня доходят таким вот печальным способом. Надеюсь, их будет немного, я буду очень стараться).
Не знаешь такой рассказ? https://nukadeti.ru/skazki/panteleev-bukva-ty

**Comment by uliss4ever** [comment]
Да... self.head указывает на None, то есть на ничто. И в обычной жизни то, чего нет, обычно так чётко не формулируется, а тут у нас это отдельная субстанция прямо... 
>А как тебе нравится математика, где в каждом уравнении по иксу стоит?!

Ага, чудесная такая математика, со всеми неизвестными) 
Но знаешь, я в начале правда не понимала, а в чём проблема-то? Запустите код, если он непонятным кажется, и посмотрите, что он будет делать. Три раза ха)

**Comment by Lex-2008** [comment]
> Запустите код, если он непонятным кажется, и посмотрите, что он будет делать. Три раза ха)

Хмм, ну как бы под микроскопом дебаггера большинство программ обычно становится понятнее. Но, конечно, имена переменных сильно помогают понять происходящее. Например, когда ты видишь классы со словами "фабрика" и "билдер" в названиях, то понимаешь, как один относится к другому, даже не глядя в код.

**Comment by uliss4ever** [comment]
Понимаешь, были времена, когда я дебаггером пользоваться не умела. Какая-то штука с зелёным тараканчиком, которой показали как пользоваться и сказали "надо", а у меня как всегда всё не получается и она только мешает.
Но эти времена прошли... настали правда, другие. Лёш, я вот смотрю на вчерашнее и мне так стыдно. Прости, что я тупила не в одиночестве с твоими объяснениями и рисунками, которых было более чем достаточно, а показывала все эти промежуточные кошмарные результаты...

**Comment by Lex-2008** [comment]
Ой, да нормально всё! "Не ошибается тот, кто ничего не делает" - не тупит тот, кто ни о чем думает! :)

Все же когда-то были новичками и делали ошибки, а теперь оглядываются и видят, какие глупости делали. И ведь это же хорошо: если ты сегодня видишь, какие глупости делала вчера - значит, сегодня ты умнее чем вчера! Супер!! Далеко не всем и не всегда удается заметить изменения в себе!

**Comment by Lex-2008** [comment]
По поводу оптимизаций и вот этих двух строчек:

```
            old_node = self._get_node_(index)
            old_node_left = self._get_node_(index-1)
```

смотри: если отбросить питонячество с итераторами, то функцию `_get_node_` для простого односвязного списка в двух словах можно описать так: "если ей дать какое-то число index, то она столько раз перейдёт от одной ноды к другой", или в коде вот так:

```
    def _get_node_(self, index):
        node = self.head
        for i in range(index):      # вот это
            node = node.next_node   # главная часть
        return node
```

Здесь главное - что какое число index передать в эту функцию, столько раз выполнится строчка `node = node.next_node`. Если index = 10, то строчка выполнится 10 раз. Если 20 - то 20. Это называется "_линейная сложность алгоритма_", и записывается `O(N)`.

`_get_node_` для двусвязного списка ведет себя примерно так же, только ещё может считать с обратной стороны, поэтому может работать быстрее, но основная тенденция сохраняется - с ростом длины списка и числа index количество операций тоже растёт.

Теперь давай посмотрим на вот эту твою картинку:
![IMG_8338](https://user-images.githubusercontent.com/32711430/110211958-20841800-7eaa-11eb-9b2d-b6ca6ba85efc.jpeg)

`old_node` указывает на ноду с вишней, `old_node_left` - с бананом. До инсёрта (верхняя часть рисунка) они были рядом друг с другом, и друг с другом они были связаны чем?

Значит, если у нас есть `old_node`, то как можно перейти к `old_node_left`, не вызывая функцию `_get_node_`?

**Comment by uliss4ever** [comment]
>До инсёрта (верхняя часть рисунка) они были рядом друг с другом, и друг с другом они были связаны чем?

Они были связаны:
old_node_left.next_node = old_node
old_node.prev_node = old_node_left
 Ааа, значит так:
old_node_left = old_node.prev_node вместо old_node_left = self._get_node_(index-1) ? Чтобы он лишний раз не бегал? Но для old_node = self._get_node_(index) вызов _get_node_ прилётся оставить, правильно же понимаю?

Я только в коммите этого не сделала, но там не все изменения. Билдеры добавила... Буду думать, но это последнее замечание Енота - ну жесть конечно...) Я бы ещё месяцок-другой коз и Бурёнок попревращала во что-нибудь, а не сразу вот это вот)

**Comment by Lex-2008** [comment]
Да, ты всё правильно понимаешь!

А вот это сможешь так же оптимизировать?
````
            new_node_l = self._get_node_(index - 1)
            new_node_r = self._get_node_(index + 1)
````

**Comment by uliss4ever** [comment]
> А вот это сможешь так же оптимизировать?
new_node_l = self._get_node_(index - 1)
new_node_r = self._get_node_(index + 1)

Вот так?
new_node_l = self._get_node_(index-1)
new_node_r = new_node_l.next_node.next_node

Да, про О большое нам говорили много на алгоритмизации. Не скажу, что я его уверенно прям везде считаю, но воочию увидела, что может быть, когда нечаянно запихнула сортировку массива в цикл фор (а мне только один раз отсортированный массив нужен был). По времени вместо 40 мс функция выполнялась 1144, этот шок забыть невозможно))

**Comment by Lex-2008** [comment]
> ````
> new_node_l = self.get_node(index-1)
> new_node_r = new_node_l.next_node.next_node
> ````

Да, можно так :)

Либо так, чтобы было симметрично:
````
old_node = self._get_node_(index)
new_node_l = old_node.prev_node
new_node_r = old_node.next_node
````

главное - что `self._get_node_` вызываем только один раз.

**Comment by uliss4ever** [comment]
Мне очень хотелось попробовать next от next написать)) Но как у тебя - красивее, конечно же. И такое чувство, что ещё немного более экономично, чем как у меня.
